### ISSUE TITLE
Use ArrayBuilder::*_check in Numba for cases of non-literal strings

### ISSUE THREAD

**Issue by jrueb** [scikit-hep/awkward#1420]
### Version of Awkward Array

1.8.0

### Description and code to reproduce

```python

import numba
import awkward as ak


builder = ak.ArrayBuilder()


@numba.njit
def build():
    builder.begin_list()
    builder.begin_record()
    for field_name in ("x", "y", "z"):
        builder.field(field_name)
        builder.integer(5)
    builder.end_record()
    builder.end_list()


build()
print(builder.snapshot())
```

Gives me the following error:

```
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
- Resolution failure for literal arguments:
wrong number or types of arguments for ArrayBuilder.field

(https://github.com/scikit-hep/awkward-1.0/blob/1.8.0/src/awkward/_connect/_numba/builder.py#L292)
- Resolution failure for non-literal arguments:
None

During: resolving callee type: BoundFunction(field for ak.ArrayBuilderType(None))
During: typing of call at numbatest.py (15)


File "numbatest.py", line 15:
def build():
    <source elided>
    for field_name in ("x", "y", "z"):
        builder.field(field_name)
        ^
```

It works if I write it like `builder.field("x")` and remove the for loop, but I need the for loop because in practice I have a long list of fields I want to add.

numba version 0.54.1

**Comment by agoose77** [comment]
Currently we only handle literal strings in the Numba implementation:
https://github.com/scikit-hep/awkward-1.0/blob/d977b04facb196a4c8584f256b36280a4e105b14/src/awkward/_connect/_numba/builder.py#L490-L502

You can work around this (with limitations) by using `numba.literal_unroll`:
```python
import numba
import awkward as ak
from numba import literal_unroll


builder = ak.ArrayBuilder()

@numba.njit
def build():
    builder.begin_list()
    builder.begin_record()
    for field_name in literal_unroll(("x", "y", "z")):
        builder.field(field_name)
        builder.integer(5)
    builder.end_record()
    builder.end_list()


build()
print(builder.snapshot())
```

Be warned, for some reason you need to import the `literal_unroll` into the global namespace (or rather, `numba.literal_unroll` [doesn't seem to work](https://github.com/numba/numba/issues/5344))

**Comment by jpivarski** [comment]
That's right, and it was on purpose, so that the Numba `builder.field` implementation can use the C++ [ArrayBuilder::field_fast](https://awkward-array.readthedocs.io/en/latest/_static/classawkward_1_1ArrayBuilder.html#a950563074bc8987b63196de7d9066041), which only checks to see if a string has the same pointer as the one for a given field, rather than checking whole strings in the loop.

I didn't know about Numba's `nb.literal_unroll`: that's a neat trick, and it certainly solves this case (in which the field names are known at compile time).

This could be a feature request for handling the case in which the string is not a literal, implemented by calling the C++ [ArrayBuilder::field_check](https://awkward-array.readthedocs.io/en/latest/_static/classawkward_1_1ArrayBuilder.html#a81883f6d746612b422a32b7990b9450f) function.

I think that would not be too difficult. Looking at a Numba-extension can be bewildering, but the change would be rather small; the person doing it wouldn't need to understand everything. I'll label this as a "good first issue," though it probably _looks_ more difficult than other "good first issues."

--------------

For symmetry, note that when iterating over an input array, the string passed to `ak.Array.__getitem__` has to be a literal string. We can handle the non-literal case for ArrayBuilder because ArrayBuilder presents itself as a single, unchanging type to Numba, but when iterating over input arrays, the type presented to Numba is specialized to the Awkward type (actually, the full Form, which is even more precise). If you could pass a runtime string to `ak.Array.__getitem__`, the Numba type of the return value would be a variant. I haven't seen much support for variants in Numba (and couldn't implement Awkward UnionArray for that reason, #174).

**Comment by jpivarski** [comment]
I changed the name because this could apply to record names as well as field names.

**Comment by asnaylor** [comment]
Tried to do something similar with ArrayBuilder:
```python
import numba as nb
import awkward as ak

@nb.njit
def numba_build(array, output):
    for evt in range(len(array)):
        output.append(array[evt])
    return output

array = ak.Array(['a', 'b', 'c'])
out = numba_build(array, ak.ArrayBuilder())
print(out.snapshot())
```
And got a similar error, must be because it's a non-literal string:
<details>
<summary>Error Traceback</summary>
<br>

```python
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
<ipython-input-42-1b6adf46919b> in <module>
     12 # array = ak.Array([1, 2, 3, 4, 5])
     13 array = ak.Array(['a', 'b', 'c'])
---> 14 out = numba_build(array, ak.ArrayBuilder())
     15 print(out.snapshot())

~/.conda/envs/LZ_Python_Kernel/lib/python3.8/site-packages/numba/core/dispatcher.py in _compile_for_args(self, *args, **kws)
    466                 e.patch_message(msg)
    467 
--> 468             error_rewrite(e, 'typing')
    469         except errors.UnsupportedError as e:
    470             # Something unsupported is present in the user code, add help info

~/.conda/envs/LZ_Python_Kernel/lib/python3.8/site-packages/numba/core/dispatcher.py in error_rewrite(e, issue_type)
    407                 raise e
    408             else:
--> 409                 raise e.with_traceback(None)
    410 
    411         argtypes = []

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
- Resolution failure for literal arguments:
unicode_type
- Resolution failure for non-literal arguments:
None

During: resolving callee type: BoundFunction(append for ak.ArrayBuilderType(None))
During: typing of call at <ipython-input-42-1b6adf46919b> (8)


File "<ipython-input-42-1b6adf46919b>", line 8:
def numba_build(array, output):
    <source elided>
    for evt in range(len(array)):
        output.append(array[evt])
        ^
```

</details>

I tried to just put in a literal string:
```python

import numba as nb
import awkward as ak

@nb.njit
def numba_build(output):
    for evt in range(3):
        output.append("x")
    return output

out = numba_build(ak.ArrayBuilder())
print(out.snapshot())
```

And i'm also seeing an error:
```python
TypingError: Failed in nopython mode pipeline (step: nopython frontend)
- Resolution failure for literal arguments:
Literal[str](x)
- Resolution failure for non-literal arguments:
unicode_type
```
<details>
<summary>Full Error Traceback</summary>
<br>

```python
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
<ipython-input-46-f6a014884374> in <module>
     10 
     11 array = ak.Array(['a', 'b', 'c'])
---> 12 out = numba_build(array, ak.ArrayBuilder())
     13 print(out.snapshot())

~/.conda/envs/LZ_Python_Kernel/lib/python3.8/site-packages/numba/core/dispatcher.py in _compile_for_args(self, *args, **kws)
    466                 e.patch_message(msg)
    467 
--> 468             error_rewrite(e, 'typing')
    469         except errors.UnsupportedError as e:
    470             # Something unsupported is present in the user code, add help info

~/.conda/envs/LZ_Python_Kernel/lib/python3.8/site-packages/numba/core/dispatcher.py in error_rewrite(e, issue_type)
    407                 raise e
    408             else:
--> 409                 raise e.with_traceback(None)
    410 
    411         argtypes = []

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
- Resolution failure for literal arguments:
Literal[str](x)
- Resolution failure for non-literal arguments:
unicode_type

During: resolving callee type: BoundFunction(append for ak.ArrayBuilderType(None))
During: typing of call at <ipython-input-46-f6a014884374> (8)


File "<ipython-input-46-f6a014884374>", line 8:
def numba_build(array, output):
    <source elided>
    for evt in range(3):
        output.append("x")
        ^
```
</details>

---

## Versions

```
awkward==1.8.0 
numba==0.55.1
python==3.8.12
```

**Comment by jpivarski** [comment]
The original issue was about literal/non-literal strings in selecting record field names, which is relevant because that's usually compiled in. (In a compiled language, the field of a struct is not quoted. In ArrayBuilder, it's only quoted because it's making a new type that didn't exist at compile-time.)

Your issue, @asnaylor, is not related to literalness, which is why it doesn't work in either case. But you're not trying to access or create a record field named "x", you're trying to add "x" as data to an array of strings. That case should absolutely work for non-literal strings, since we'd want to fill an array with different strings in different array elements. You're using [ArrayBuilder.append](https://awkward-array.readthedocs.io/en/latest/_auto/ak.ArrayBuilder.html#ak-arraybuilder-append), which is a convenience method that selects between different methods based on the type of its argument (including string). I bet I didn't include that among the overloads in Numba, that I simply forgot that `append` includes strings as well as `integer`, `real`, etc.

Yes, that's what happened:

https://github.com/scikit-hep/awkward-1.0/blob/ec9eec9aa5fae80dac44c300c6a48e3766afd1cc/src/awkward/_connect/_numba/builder.py#L303-L335

The Numba implementation of `append` accepts `numba.types.Boolean`, `numba.types.Integer`, and `numba.types.Float`, but not the others.

Your immediate problem can be fixed by just using the [ArrayBuilder.string](https://awkward-array.readthedocs.io/en/latest/_auto/ak.ArrayBuilder.html#ak-arraybuilder-string) method, but I really need to make the Numba implementation consistent with the Python!

**Comment by agoose77** [comment]
We also don't seem to support `ArrayBuilder.string` either!

**Comment by asnaylor** [comment]
Ah that makes sense about the literalness. Yeah when i try with `.string` instead of `append` for the ArrayBuilder in the numba function i still get a `TypingError`: `Unknown attribute 'string' of type ak.ArrayBuilderType(None)`.

**Comment by jpivarski** [comment]
I'm astonished: it's weird that I would just forget that. But it does explain why `append` isn't overloaded to cover it. Probably the feature was added later and the ArrayBuilder-in-Numba implementation is just behind the times?

```python
>>> import numba as nb
>>> import awkward as ak
>>> @nb.njit
... def add_a_string(builder, string):
...     builder.string(string)
...     return builder
... 
>>> builder = add_a_string(ak.ArrayBuilder(), "hello")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jpivarski/mambaforge/lib/python3.9/site-packages/numba/core/dispatcher.py", line 468, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/home/jpivarski/mambaforge/lib/python3.9/site-packages/numba/core/dispatcher.py", line 409, in error_rewrite
    raise e.with_traceback(None)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Unknown attribute 'string' of type ak.ArrayBuilderType(None)

File "<stdin>", line 3:
<source missing, REPL/exec in use?>

During: typing of get attribute at <stdin> (3)

File "<stdin>", line 3:
<source missing, REPL/exec in use?>
```

Indeed: it's missing. Also, I don't see it in either the v1 or the v2 implementation.

This should be a separate issue: ArrayBuilder-in-Numba is missing `string` and `bytestring`, and possibly other new methods (`complex`? `datetime`?). Fortunately, we know what the interface is _supposed_ to be: it's supposed to be the same as outside of Numba, the ArrayBuilder Python interface.

(BTW: ArrayBuilder-in-Numba's lack of context managers `list`, `record`, and `tuple` is known because it's waiting on context managers as a feature from Numba, but it sounds like this will be implemented soon, and Awkward Array will be a first use-case.)

**Comment by jpivarski** [comment]
Oh, and if you need an interim solution for filling strings in Numba, Awkward strings are really just lists of `np.uint8` with parameters attached: `__array__: "char"` for the NumpyArray and `__array__: "string"` for the ListOffsetArray. If you build it from two NumPy arrays (`np.uint8` characters and `np.int64` offsets) that you later construct into an Awkward Array, your implementation will be faster than ArrayBuilder will be, when the feature does get added.

```python
>>> ak.Array(["These", "are", "some", "strings."]).layout
<ListOffsetArray64>
    <parameters>
        <param key="__array__">"string"</param>
    </parameters>
    <offsets><Index64 i="[0 5 8 12 20]" offset="0" length="5" at="0x55c407ed2e10"/></offsets>
    <content><NumpyArray format="B" shape="20" data="84 104 101 115 101 ... 105 110 103 115 46" at="0x55c407bbd870">
        <parameters>
            <param key="__array__">"char"</param>
        </parameters>
    </NumpyArray></content>
</ListOffsetArray64>
```
